# Advent of Code, 12th Dec 2024

## Garden Groups - Part 1

Why not search for the Chief Historian near the gardener and his massive farm? There's plenty of food, so The Historians grab something to eat while they search.

You're about to settle near a complex arrangement of garden plots when some Elves ask if you can lend a hand. They'd like to set up fences around each region of garden plots, but they can't figure out how much fence they need to order or how much it will cost. They hand you a map (your puzzle input) of the garden plots.

Each garden plot grows only a single type of plant and is indicated by a single letter on your map. When multiple garden plots are growing the same type of plant and are touching (horizontally or vertically), they form a **region**. For example:

```
AAAA
BBCD
BBCC
EEEC
```

This 4x4 arrangement includes garden plots growing five different types of plants (labeled A, B, C, D, and E), each grouped into their own region.

In order to accurately calculate the cost of the fence around a single region, you need to know that region's **area** and **perimeter**.

The **area** of a region is simply the number of garden plots the region contains. The above map's type A, B, and C plants are each in a region of area 4. The type E plants are in a region of area 3; the type D plants are in a region of area 1.

Each garden plot is a square and so has **four sides**. The **perimeter** of a region is the number of sides of garden plots in the region that do not touch another garden plot in the same region. The type A and C plants are each in a region with perimeter 10. The type B and E plants are each in a region with perimeter 8. The lone D plot forms its own region with perimeter 4.

Visually indicating the sides of plots in each region that contribute to the perimeter using - and |, the above map's regions' perimeters are measured as follows:

```
+-+-+-+-+
|A A A A|
+-+-+-+-+     +-+
              |D|
+-+-+   +-+   +-+
|B B|   |C|
+   +   + +-+
|B B|   |C C|
+-+-+   +-+ +
          |C|
+-+-+-+   +-+
|E E E|
+-+-+-+
```

Plants of the same type can appear in multiple separate regions, and regions can even appear within other regions. For example:

```
OOOOO
OXOXO
OOOOO
OXOXO
OOOOO
```

The above map contains **five** regions, one containing all of the O garden plots, and the other four each containing a single X plot.

The four X regions each have area 1 and perimeter 4. The region containing 21 type O plants is more complicated; in addition to its outer edge contributing a perimeter of 20, its boundary with each X region contributes an additional 4 to its perimeter, for a total perimeter of 36.

Due to "modern" business practices, the **price** of fence required for a region is found by **multiplying** that region's area by its perimeter. The **total price** of fencing all regions on a map is found by adding together the price of fence for every region on the map.

In the first example, region A has price 4 * 10 = 40, region B has price 4 * 8 = 32, region C has price 4 * 10 = 40, region D has price 1 * 4 = 4, and region E has price 3 * 8 = 24. So, the total price for the first example is **140**.

In the second example, the region with all of the O plants has price 21 * 36 = 756, and each of the four smaller X regions has price 1 * 4 = 4, for a total price of **772** (756 + 4 + 4 + 4 + 4).

Here's a larger example:

```
RRRRIICCFF
RRRRIICCCF
VVRRRCCFFF
VVRCCCJFFF
VVVVCJJCFE
VVIVCCJJEE
VVIIICJJEE
MIIIIIJJEE
MIIISIJEEE
MMMISSJEEE
```

It contains:

* A region of R plants with price 12 * 18 = 216.
* A region of I plants with price 4 * 8 = 32.
* A region of C plants with price 14 * 28 = 392.
* A region of F plants with price 10 * 18 = 180.
* A region of V plants with price 13 * 20 = 260.
* A region of J plants with price 11 * 20 = 220.
* A region of C plants with price 1 * 4 = 4.
* A region of E plants with price 13 * 18 = 234.
* A region of I plants with price 14 * 22 = 308.
* A region of M plants with price 5 * 12 = 60.
* A region of S plants with price 3 * 8 = 24.

So, it has a total price of **1930**.

**What is the total price of fencing all regions on your map?**

### Read in the data

In [66]:
#!F#
let inputFilePath = "aoc-2024-12-12-puzzle-input.txt"

Read the lines from the file.

In [67]:
open System.IO

let lines: string list = File.ReadAllLines(inputFilePath) |> Array.toList
lines.Length

140

In [68]:
let minLineLength = lines |> List.map (fun line -> line.Length) |> List.min
let maxLineLength = lines |> List.map (fun line -> line.Length) |> List.max
(minLineLength, maxLineLength)

(140, 140) Item1 140 Item2 140

So the input is a 140*140 square of characters.

### Solve the problem

In [69]:
let N = lines.Length
N

140

In [70]:
let gardenMap: char list list = lines |> List.rev |> List.map (fun line -> line |> Seq.toList)
gardenMap[0]

[ C, C, C, C, C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (values) [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V ... (more) ] HeadOrDefault C TailOrNull [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] Head C Tail [ C, C, C, C, C, C, R, W, W, R, R, V, V, V, V, V, V, V, V, V ... (more) ] (value

In [71]:
type Region = { plantType: char; points: (int*int) Set }

let area (region: Region): int = region.points |> Set.count

let isEdgePoint (region: Region) (point: int*int): bool =
    match point with
    | (col, row) when col = 0 || col = N-1 || row = 0 || row = N-1 -> true
    | (col, row) when not (
            (region.points |> Set.contains (col-1, row)) &&
            (region.points |> Set.contains (col+1, row)) &&
            (region.points |> Set.contains (col, row-1)) &&
            (region.points |> Set.contains (col, row+1))
        ) -> true
    | _ -> false

let countExposedEdges (region: Region) (point: int*int): int =
    let (col, row) = point
    [
        col = 0 || not (region.points |> Set.contains (col-1, row)) ;
        col = N-1 || not (region.points |> Set.contains (col+1, row)) ;
        row = 0 || not (region.points |> Set.contains (col, row-1)) ;
        row = N-1 || not (region.points |> Set.contains (col, row+1))
    ] |> List.filter (fun boolVal -> boolVal) |> List.length

let perimeter (region: Region): int =
    region.points |> Set.toSeq |> Seq.map (fun point -> countExposedEdges region point) |> Seq.sum

So, having done all of the above, how can we isolate the various regions in the garden?  A possible solution is:
* Create a region for each point in the garden, and include in that region all neighbouring points that have the same plant type (excluding diagonal neighbours).
* Unify any regions that have the same plant type and overlap (i.e. share any of the same points).

In [72]:
let neighbours (gardenMap: char list list) (plantType: char) (point: int*int): (int*int) Set =
    let (col, row) = point
    (seq {
        if (col > 0 && gardenMap[row][col-1] = plantType) then yield (col-1, row)
        if (col < (N-1) && gardenMap[row][col+1] = plantType) then yield (col+1, row)
        if (row > 0 && gardenMap[row-1][col] = plantType) then yield (col, row-1)
        if (row < (N-1) && gardenMap[row+1][col] = plantType) then yield (col, row+1)
    }) |> Set.ofSeq

let rec unifyRegions (regions: Region list): Region list =
    match regions with
    | [] -> []
    | head::[] -> regions
    | head::tail ->
        let (unionTail, remainder) = tail |> List.partition (fun reg -> reg.plantType = head.plantType && not (Set.intersect head.points reg.points).IsEmpty)
        match unionTail with
        | [] -> head::(unifyRegions tail)
        | _ ->
            let allPoints = Set(seq{ for pt in head.points do yield pt ; for reg in unionTail do for pt in reg.points do yield pt} )
            unifyRegions ({ plantType = head.plantType; points = allPoints }::remainder)

let buildRegions (gardenMap: char list list): Region list =
    let startRegions =
        (seq {
            for col in 0..(N-1) do
                for row in 0..(N-1) do
                    let plantType = gardenMap[row][col]
                    yield { plantType = plantType; points = (neighbours gardenMap plantType (col,row)) |> Set.add (col, row) }
        }) |> Seq.toList
    startRegions |> unifyRegions
    

In [73]:
let gardenRegions = buildRegions gardenMap
gardenRegions

[ { plantType = 'C'
 points =
 set
 [(0, 0); (0, 1); (0, 2); (0, 3); (0, 4); (0, 5); (0, 6); (0, 7); (0, 8);
 ...] }, { plantType = 'I'
 points =
 set
 [(0, 11); (0, 12); (0, 14); (0, 15); (0, 16); (0, 17); (0, 18); (0, 19);
 (0, 20); ...] }, { plantType = 'Q'
 points = set [(0, 13); (1, 13); (1, 14)] }, { plantType = 'Z'
 points = set [(0, 23)] }, { plantType = 'G'
 points =
 set
 [(0, 24); (0, 25); (0, 26); (0, 29); (1, 23); (1, 24); (1, 25); (1, 26);
 (1, 27); ...] }, { plantType = 'N'
 points = set [(0, 27); (0, 28)] }, { plantType = 'K'
 points = set [(0, 30); (0, 31); (0, 32); (1, 30); (1, 31); (1, 32)] }, { plantType = 'P'
 points =
 set
 [(0, 33); (0, 34); (0, 35); (0, 36); (0, 37); (0, 38); (0, 39); (0, 40);
 (0, 41); ...] }, { plantType = 'E'
 points = set [(0, 43); (0, 44); (0, 45)] }, { plantType = 'A'
 points =
 set
 [(0, 46); (0, 47); (0, 48); (0, 49); (0, 50); (0, 51); (0, 52); (0, 53);
 (0, 54); ...] }, { plantType = 'M'
 points =
 set
 [(0, 64); (0, 65); (0, 66); (0, 67); (0, 68); (0, 69); (1, 65); (1, 66);
 (1, 67); ...] }, { plantType = 'U'
 points =
 set
 [(0, 70); (0, 71); (0, 72); (0, 73); (0, 74); (0, 75); (0, 76); (0, 77);
 (0, 78); ...] }, { plantType = 'M'
 points =
 set
 [(0, 81); (0, 82); (0, 83); (1, 81); (1, 83); (2, 81); (2, 83); (3, 76);
 (3, 77); ...] }, { plantType = 'I'
 points =
 set
 [(0, 84); (0, 85); (0, 86); (1, 84); (1, 85); (1, 86); (2, 84); (2, 85);
 (2, 86); ...] }, { plantType = 'Y'
 points =
 set
 [(0, 87); (0, 88); (0, 89); (0, 90); (0, 91); (0, 92); (0, 93); (0, 94);
 (0, 95); ...] }, { plantType = 'I'
 points =
 set
 [(0, 96); (0, 98); (0, 99); (0, 100); (0, 101); (0, 102); (0, 103);
 (0, 104); (0, 105); ...] }, { plantType = 'Q'
 points = set [(0, 97)] }, { plantType = 'L'
 points = set [(0, 120); (1, 120)] }, { plantType = 'Y'
 points =
 set
 [(0, 121); (0, 122); (0, 123); (0, 124); (0, 125); (0, 126); (0, 127);
 (0, 128); (0, 129); ...] }, { plantType = 'W'
 points =
 set
 [(0, 131); (0, 132); (0, 133); (0, 134); (0, 135); (0, 136); (0, 137);
 (0, 138); (0, 139); ...] } ... (more) ] HeadOrDefault { plantType = 'C'\n points =\n set\n [(0, 0); (0, 1); (0, 2); (0, 3); (0, 4); (0, 5); (0, 6); (0, 7); (0, 8);\n ...] } plantType C points [ (0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8) ... (128 more) ] Count 148 (values) index value 0 (0, 0) Item1 0 Item2 0 1 (0, 1) Item1 0 Item2 1 2 (0, 2) Item1 0 Item2 2 3 (0, 3) Item1 0 Item2 3 4 (0, 4) Item1 0 Item2 4 5 (0, 5) Item1 0 Item2 5 6 (0, 6) Item1 0 Item2 6 7 (0, 7) Item1 0 Item2 7 8 (0, 8) Item1 0 Item2 8 9 (0, 9) Item1 0 Item2 9 10 (0, 10) Item1 0 Item2 10 11 (1, 0) Item1 1 Item2 0 12 (1, 1) Item1 1 Item2 1 13 (1, 2) Item1 1 Item2 2 14 (1, 3) Item1 1 Item2 3 15 (1, 4) Item1 1 Item2 4 16 (1, 5) Item1 1 Item2 5 17 (1, 6) Item1 1 Item2 6 18 (1, 7) Item1 1 Item2 7 19 (1, 8) Item1 1 Item2 8 ... (more) TailOrNull [ { plantType = 'I'
 points =
 set
 [(0, 11); (0, 12); (0, 14); (0, 15); (0, 16); (0, 17); (0, 18); (0, 19);
 (0, 20); ...] }, { plantType = 'Q'
 points = set [(0, 13); (1, 13); (1, 14)] }, { plantType = 'Z'
 points = set [(0, 23)] }, { plantType = 'G'
 points =
 set
 [(0, 24); (0, 25); (0, 26); (0, 29); (1, 23); (1, 24); (1, 25); (1, 26);
 (1, 27); ...] }, { plantType = 'N'
 points = set [(0, 27); (0, 28)] }, { plantType = 'K'
 points = set [(0, 30); (0, 31); (0, 32); (1, 30); (1, 31); (1, 32)] }, { plantType = 'P'
 points =
 set
 [(0, 33); (0, 34); (0, 35); (0, 36); (0, 37); (0, 38); (0, 39); (0, 40);
 (0, 41); ...] }, { plantType = 'E'
 points = set [(0, 43); (0, 44); (0, 45)] }, { plantType = 'A'
 points =
 set
 [(0, 46); (0, 47); (0, 48); (0, 49); (0, 50); (0, 51); (0, 52); (0, 53);
 (0, 54); ...] }, { plantType = 'M'
 points =
 set
 [(0, 64); (0, 65); (0, 66); (0, 67); (0, 68); (0, 69); (1, 65); (1, 66);
 (1, 67); ...] }, { plantType = 'U'
 points =
 set
 [(0, 70); (0, 71); (0, 72); (0, 73); (0, 74); (0, 75); (0,

In [74]:
let totalPrice = gardenRegions |> List.map (fun region -> (area region) * (perimeter region)) |> List.sum
totalPrice

1550156

## Garden Groups - Part 2

Fortunately, the Elves are trying to order so much fence that they qualify for a **bulk discount**!

Under the bulk discount, instead of using the perimeter to calculate the price, you need to use the **number of sides** each region has. Each straight section of fence counts as a side, regardless of how long it is.

Consider this example again:

```
AAAA
BBCD
BBCC
EEEC
```

The region containing type A plants has 4 sides, as does each of the regions containing plants of type B, D, and E. However, the more complex region containing the plants of type C has 8 sides!

Using the new method of calculating the per-region price by multiplying the region's area by its number of sides, regions A through E have prices 16, 16, 32, 4, and 12, respectively, for a total price of **80**.

The second example above (full of type X and O plants) would have a total price of **436**.

Here's a map that includes an E-shaped region full of type E plants:

```
EEEEE
EXXXX
EEEEE
EXXXX
EEEEE
```

The E-shaped region has an area of 17 and 12 sides for a price of 204. Including the two regions full of type X plants, this map has a total price of **236**.

This map has a total price of **368**:

```
AAAAAA
AAABBA
AAABBA
ABBAAA
ABBAAA
AAAAAA
```

It includes two regions full of type B plants (each with 4 sides) and a single region full of type A plants (with 4 sides on the outside and 8 more sides on the inside, a total of 12 sides). Be especially careful when counting the fence around regions like the one full of type A plants; in particular, each section of fence has an in-side and an out-side, so the fence does not connect across the middle of the region (where the two B regions touch diagonally). (The Elves would have used the Möbius Fencing Company instead, but their contract terms were too one-sided.)

The larger example from before now has the following updated prices:

* A region of R plants with price 12 * 10 = 120.
* A region of I plants with price 4 * 4 = 16.
* A region of C plants with price 14 * 22 = 308.
* A region of F plants with price 10 * 12 = 120.
* A region of V plants with price 13 * 10 = 130.
* A region of J plants with price 11 * 12 = 132.
* A region of C plants with price 1 * 4 = 4.
* A region of E plants with price 13 * 8 = 104.
* A region of I plants with price 14 * 16 = 224.
* A region of M plants with price 5 * 6 = 30.
* A region of S plants with price 3 * 6 = 18.

Adding these together produces its new total price of **1206**.

**What is the new total price of fencing all regions on your map?**

### Solve the problem

So how can we calculate the number of sides of each region?  It won't be as simple as the trick we used to calculate the perimeter.  We can
* Isolate the points that are on the boundary (edge) of each region.
* Calculate the edges for each of those edge points.
  * While point indices run from 0..N-1 in each direction, edge indices run 0..N in each direction.
* Sort edges into order, then take the first edge as the start of a side, look for other edges that complete that side, etc.
* **NOTE** - after some experimentation, it was discovered that a 'side' is more than just a contiguous set of edges in the same direction, it *also* has to have the points (garden blocks) on the *same side* of the whole edge.  This changes the result that you would otherwise get.

In [75]:
type EdgeDirection = Horizontal | Vertical

type Edge = { point: int*int; direction: EdgeDirection; edgeStart: int*int; edgeEnd: int*int }

let edges (region: Region) (point: int*int): Edge Set =
    if (isEdgePoint region point) then
        let (col, row) = point
        (seq {
            if (col = 0) then yield {point = point; direction = Vertical; edgeStart = (0, row); edgeEnd = (0, row+1)}
            if (col = (N-1)) then yield {point = point; direction = Vertical; edgeStart = (N, row); edgeEnd = (N, row+1)}
            if (row = 0) then yield {point = point; direction = Horizontal; edgeStart = (col, 0); edgeEnd = (col+1, 0)}
            if (row = (N-1)) then yield {point = point; direction = Horizontal; edgeStart = (col, N); edgeEnd = (col+1, N)}
            if (col > 0 && not (region.points |> Set.contains (col-1, row))) then yield {point = point; direction = Vertical; edgeStart = (col, row); edgeEnd = (col, row+1)}
            if (col < N-1 && not (region.points |> Set.contains (col+1, row))) then yield {point = point; direction = Vertical; edgeStart = (col+1, row); edgeEnd = (col+1, row+1)}
            if (row > 0 && not (region.points |> Set.contains (col, row-1))) then yield {point = point; direction = Horizontal; edgeStart = (col, row); edgeEnd = (col+1, row)}
            if (row < N-1 && not (region.points |> Set.contains (col, row+1))) then yield {point = point; direction = Horizontal; edgeStart = (col, row+1); edgeEnd = (col+1, row+1)}
        } |> Set.ofSeq)
    else Set.empty

In [79]:
let sortEdgesIntoSides (edges: Edge list): Edge list list =
    let horizontalEdges = edges |> List.filter (fun edge -> edge.direction = Horizontal) |> List.sortBy (fun edge -> edge.edgeStart)
    let verticalEdges = edges |> List.filter (fun edge -> edge.direction = Vertical) |> List.sortBy  (fun edge -> edge.edgeStart)
    let rec collectEdgesIntoSides (sides: Edge list list) (edges: Edge list): Edge list list =
        match edges with
        | [] -> sides
        | head::tail ->
            match sides with
            | [] -> collectEdgesIntoSides [[head]] tail
            | side::rest ->
                let sideLastEdge = side |> List.last
                let nextEdgeList, otherEdgesList = edges |> List.partition (fun edge ->
                    edge.edgeStart = sideLastEdge.edgeEnd && (
                        (edge.direction = Vertical && (fst edge.point) = (fst sideLastEdge.point)) ||
                        (edge.direction = Horizontal && (snd edge.point) = (snd sideLastEdge.point))
                    )
                )
                match nextEdgeList with
                | [] ->
                    let sides = collectEdgesIntoSides rest edges
                    side::sides
                | _ ->
                    collectEdgesIntoSides ((side @ nextEdgeList)::rest) otherEdgesList
    (seq {
        yield! (collectEdgesIntoSides [] horizontalEdges)
        yield! (collectEdgesIntoSides [] verticalEdges)
    } |> Seq.toList)

let sides (region: Region) =
    let edgePoints = region.points |> Set.filter (fun point -> isEdgePoint region point)
    let regionEdgeList = edgePoints |> Set.map (fun point -> edges region point) |> Set.toList |> List.fold Set.union Set.empty |> Set.toList
    let sides = sortEdgesIntoSides regionEdgeList
    sides

let sideCount (region: Region) = sides region |> List.length

In [77]:
gardenRegions |> List.map (fun region -> (region.plantType, area region, sideCount region))

[ (C, 148, 76), (I, 82, 38), (Q, 3, 6), (Z, 1, 4), (G, 56, 28), (N, 2, 4), (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32) ... (more) ] HeadOrDefault (C, 148, 76) Item1 C Item2 148 Item3 76 TailOrNull [ (I, 82, 38), (Q, 3, 6), (Z, 1, 4), (G, 56, 28), (N, 2, 4), (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14) ... (more) ] HeadOrDefault (I, 82, 38) Item1 I Item2 82 Item3 38 TailOrNull [ (Q, 3, 6), (Z, 1, 4), (G, 56, 28), (N, 2, 4), (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4) ... (more) ] HeadOrDefault (Q, 3, 6) Item1 Q Item2 3 Item3 6 TailOrNull [ (Z, 1, 4), (G, 56, 28), (N, 2, 4), (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4) ... (more) ] HeadOrDefault (Z, 1, 4) Item1 Z Item2 1 Item3 4 TailOrNull [ (G, 56, 28), (N, 2, 4), (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4), (V, 59, 36) ... (more) ] HeadOrDefault (G, 56, 28) Item1 G Item2 56 Item3 28 TailOrNull [ (N, 2, 4), (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4), (V, 59, 36), (G, 2, 4) ... (more) ] HeadOrDefault (N, 2, 4) TailOrNull [ (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4), (V, 59, 36), (G, 2, 4), (T, 2, 4) ... (more) ] Head (N, 2, 4) Tail [ (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4), (V, 59, 36), (G, 2, 4), (T, 2, 4) ... (more) ] (values) index value 0 (N, 2, 4) 1 (K, 6, 4) 2 (P, 107, 46) 3 (E, 3, 4) 4 (A, 61, 20) 5 (M, 47, 26) 6 (U, 36, 16) 7 (M, 74, 42) 8 (I, 107, 44) 9 (Y, 65, 28) 10 (I, 222, 44) 11 (Q, 1, 4) 12 (L, 2, 4) 13 (Y, 87, 38) 14 (W, 87, 32) 15 (Z, 268, 14) 16 (J, 2, 4) 17 (S, 1, 4) 18 (V, 59, 36) 19 (G, 2, 4) ... (more) Head (G, 56, 28) Item1 G Item2 56 Item3 28 Tail [ (N, 2, 4), (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4), (V, 59, 36), (G, 2, 4) ... (more) ] HeadOrDefault (N, 2, 4) TailOrNull [ (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4), (V, 59, 36), (G, 2, 4), (T, 2, 4) ... (more) ] Head (N, 2, 4) Tail [ (K, 6, 4), (P, 107, 46), (E, 3, 4), (A, 61, 20), (M, 47, 26), (U, 36, 16), (M, 74, 42), (I, 107, 44), (Y, 65, 28), (I, 222, 44), (Q, 1, 4), (L, 2, 4), (Y, 87, 38), (W, 87, 32), (Z, 268, 14), (J, 2, 4), (S, 1, 4), (V, 59, 36), (G, 2, 4), (T, 2, 4) ... (more) ] (values) index value 0 (N, 2, 4) 1 (K, 6, 4) 2 (P, 107, 46) 3 (E, 3, 4) 4 (A, 61, 20) 5 (M, 47, 26) 6 (U, 36, 16) 7 (M, 74, 42) 8 (I, 107, 44) 9 (Y, 65, 28) 10 (I, 222, 44) 11 (Q, 1, 4) 12 (L, 2, 4) 13 (Y, 87, 38) 14 (W, 87, 32) 15 (Z, 268, 14) 16 (J, 2, 4) 17 (S, 1, 4) 18 (V, 59, 36) 19 (G, 2, 4) ... (more) (values) index value 0 (G

In [78]:
let reducedPrice = gardenRegions |> List.map (fun region -> (area region) * (sideCount region)) |> List.sum
reducedPrice

946084